rfcs/react-components/convergence/custom-styling.md

```diff
@@ -0,0 +1,510 @@
+  style block completely.
+
+### 🤔 Centralized and singular custom styling method
+
+It would be nice to have a single method on context that could apply for all components. This proves difficult
+with each component having its own props and state types.
+
+Another approach would be to define single context value type in the react-shared-contexts package that had
+a method for each component, but that causes a circular dependency problem for components.
+
+There may some typescript trickery that would allow for a single context value type, but having each component define
+the context type it will inspect for custom styling follows SOLID dependency inversion.
+
+## Option A2: Add global custom styling hook
+
+A global custom styling hook would be defined. It takes a component kind and state.
+
+> We experimented with several ways to try and make the method typesafe, but it proved
+> either overly complex, forced conditional calls that would make hooks unstable,
+> led to cirtcular references between packages, or restricted custom styling to the point
+> of lower value as an escape hatch.
+
+```tsx
+export type FluentStyleCustomizer = {
+  useCustomStyles_unstable: (kind: string, state: any) => void;
+};
+
+export const fuiCustomizer: FuiCustomizer = {
+  useCustomStyles_unstable: () => {},
+};
+```
+
+Components would call the custom styling hook immediately after their own styling hook.
+
+> The kind passed to the custom style hook can be specific to library (e.g. 'FluentUI.Button').
+
+```tsx
+export const SomeComponent: ForwardRefComponent<SomeComponentProps> = React.forwardRef((props, ref) => {
+  const state = useSomeComponent_unstable(props, ref);
+
+  useSomeComponentStyles_unstable(state);
+
+  fuiCustomizer.useCustomStyles_unstable('SomeComponent', state);
+
+  return renderSomeComponent_unstable(state);
+  // Casting is required due to lack of distributive union to support unions on @types/react
+}) as ForwardRefComponent<SomeComponentProps>;
+```
+
+Callers can customize styles by replacing the default useCustomStyles_unstable hook.
+
+> Callers are encouraged to use Griffel to define styles.
+> They can use makeStyles and mergeClasses to default, override, or replace
+> The classes applied by the component style hook.
+> A good implementation will switch to typesafe methods.
+
+```ts
+const useSomeComponentCustomStyles = (state: SomeComponentState) => {
+  const styles = useStyles();
+
+  if (state.active) {
+  }
+  state.root.className = mergeClasses(
+    state.root.className,
+    styles.root,
+    state.size === 'small' && styles.small,
+    state.size === 'medium' && styles.medium,
+    state.size === 'large' && styles.large,
+  );
+};
+
+fuiCustomizer.useCustomStyles_unstable = (kind: string, state: any) => {
+  switch (kind) {
+    case 'SomeComponent':
+      useSomeComponentCustomStyles(<SomeComponentState>state);
+      break;
+    // case ...
+  }
+};
```

Comment:
I don't think it's scalable, imagine that that you to customize styles for 10 components.

Can't we simply do following?

```
export type FluentStyleCustomizer = {
  useCustomButtonStyles: (state: unknown) => void;
  useCustomSelectStyles: (state: unknown) => void;
  useCustomXStyles: (state: unknown) => void;
};
```

It's the same that is done a switch anyway. I also suggest to use `unknown`, so via type cast consumers can get proper types.

Reply:
We could.  I was trying to avoid needing to modify this every time we introduce a component.  Following the open for extension, closed for modification principle.

Reply:
What Shift proposes would also mean the component would know whether there are any overrides present or not, allowing it for example to avoid calling the component styles hook.

Reply:
@miroslavstastny - we want to avoid conditionally calling a hook though right?  I was thinking we would initially have the global be a set of no-op implementations.

Reply:
Updated OptionA2 with Shift's suggested approach.  I like it better!